pkg/operators/wasm/datasource.go

```diff
@@ -0,0 +1,486 @@
+		i.logger.Warnf("failed to register datasource: %v", err)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(ds))
+}
+
+// getDataSource returns a data source by its name.
+// Params:
+// - stack[0] is the name of the datasource (string encoded)
+// Return value:
+// - DataSource handle
+func (i *wasmOperatorInstance) getDataSource(ctx context.Context, m wapi.Module, stack []uint64) {
+	dsName, err := stringFromStack(m, stack[0])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	ds := i.gadgetCtx.GetDataSources()[dsName]
+	if ds == nil {
+		i.logger.Debugf("datasource not found %q", dsName)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(ds))
+}
+
+// dataSourceGetField returns a handle to a data source.
+// Params:
+// - stack[0]: DataSource handle
+// - stack[1]: Field name
+// Return value:
+// - Field handle
+func (i *wasmOperatorInstance) dataSourceGetField(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("DataSource handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	fieldName, err := stringFromStack(m, stack[1])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	acc := ds.GetField(fieldName)
+	stack[0] = wapi.EncodeU32(i.addHandle(acc))
+}
+
+// dataSourceAddField add a field to the data source and returns its handle.
+// Params:
+// - stack[0]: DataSource handle
+// - stack[1]: Field name
+// - stack[2]: Field kind
+// Return value:
+// - Field handle
+func (i *wasmOperatorInstance) dataSourceAddField(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("DataSource handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	fieldName, err := stringFromStack(m, stack[1])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	acc, err := ds.AddField(fieldName, api.Kind(stack[2]))
+	if err != nil {
+		i.logger.Warnf("adding field %q to datasource %q: %v", fieldName, ds.Name(), err)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(acc))
+}
+
+type subscriptionType uint32
+
+const (
+	subscriptionTypeData   subscriptionType = 0
+	subscriptionTypeArray  subscriptionType = 1
+	subscriptionTypePacket subscriptionType = 2
+)
+
+// dataSourceSubscribe subscribes to the datasource.
+// Params:
+// - stack[0]: DataSource handle
+// - stack[1]: Type (0: Data, 1: Array, 2: Packet)
+// - stack[2]: Priority
+// - stack[3]: Callback ID
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) dataSourceSubscribe(ctx context.Context, m wapi.Module, stack []uint64) {
+	if i.dsCallback == nil {
+		i.logger.Warnf("wasm module doesn't export dsCallback")
+		stack[0] = 1
+		return
+	}
```

Comment:
Exactly, the wasm operator can add new fields, etc. without having to subscribe, so that's why this is optional. We can check later on if we wan to make this mandatory.